test/jdk/java/text/Format/NumberFormat/Bug4838107.java

```diff
@@ -24,228 +24,256 @@
+        );
     }
 
-    static void test(Number num, String str) {
+    // Check that the formatted value is correct and also check that
+    // it can be round-tripped via parse() and format()
+    private static void test(Number num, String str, DecimalFormat df) {
         String formatted = df.format(num);
-        if (!formatted.equals(str)) {
-            err = true;
-            System.err.println("    DecimalFormat format(" +
-                               num.getClass().getName() +
-                               ") error: \n\tnumber: " + num +
-                               "\n\tminus sign: " + dfs.getMinusSign() +
-                               "\n\tgot:        " + formatted +
-                               "\n\texpected:   " + str);
-            return;
-        }
+        assertEquals(str, formatted, String.format("DecimalFormat format(%s) " +
+                "Error: number: %s, minus sign: %s", num.getClass().getName(), num, df.getDecimalFormatSymbols().getMinusSign()));
 
         if (num instanceof BigDecimal || num instanceof BigInteger) {
             df.setParseBigDecimal(true);
         }
+        testRoundTrip(formatted, str, num, df);
+    }
+
+    // Test that a parsed value can be round-tripped via format() and parse()
+    private static void testRoundTrip(String formatted, String str,
+                                      Number num, DecimalFormat df) {
         Number parsed1 = null, parsed2 = null;
         try {
             parsed1 = df.parse(formatted);
             formatted = df.format(parsed1);
             parsed2 = df.parse(formatted);
-            if (!parsed1.equals(parsed2)) {
-                err = true;
-                System.err.println("    DecimalFormat roundtrip parse(" +
-                                   num.getClass().getName() +
-                                   ") error: \n\toriginal number:  " + str +
-                                   "\n\tparsed number:    " + parsed1 +
-                                   "  (" + parsed1.getClass().getName() + ")" +
-                                   "\n\tformatted number: " + formatted +
-                                   "\n\tre-parsed number: " + parsed2 +
-                                   "  (" + parsed2.getClass().getName() + ")" +
-                                   "\n\tminus sign: " + dfs.getMinusSign());
-            }
+            assertEquals(parsed2, parsed1, "DecimalFormat round trip parse(" +
+                    num.getClass().getName() +
+                    ") error: \n\toriginal number:  " + str +
+                    "\n\tparsed number:    " + parsed1 +
+                    "  (" + parsed1.getClass().getName() + ")" +
+                    "\n\tformatted number: " + formatted +
+                    "\n\tre-parsed number: " + parsed2 +
+                    "  (" + parsed2.getClass().getName() + ")" +
+                    "\n\tminus sign: " + df.getDecimalFormatSymbols().getMinusSign());
```

Comment:
Could use text block to get rid of `\n`/`\t`s, with `.formatted()`. Same for the following one.

Reply:
Adjusted.